mds: add an asok command to dump export states

zhsgao opened the pull request:
Task to export subtree may be blocked, use this command to find out what's going on.

Fixes: https://tracker.ceph.com/issues/58835
Signed-off-by: Zhansong Gao <zhsgao@hotmail.com>

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

vshankar commented:
Nice idea. Will have a look at the changes.

dparmar18 commented:
@zhsgao would be great to have a tracker for this

zhsgao commented:
> @zhsgao would be great to have a tracker for this

OK, I have added it.

vshankar commented on src/mds/Migrator.cc:
```diff
@@ -1391,18 +1391,18 @@ void Migrator::export_frozen(CDir *dir, uint64_t tid)
 
-  set<client_t> export_client_set;
-  get_export_client_set(dir, export_client_set);
+  ceph_assert(it->second.export_client_set.empty());
```
Although this check looks to be safe, is there a case where this isn't valid (like dispatch_export_dir() getting called multiple times for a CDir)?

dparmar18 commented on qa/tasks/cephfs/test_dump_export_states.py:
```diff
@@ -0,0 +1,95 @@
+    EXPORT_STATES = ['locking', 'discovering', 'freezing', 'prepping', 'warning', 'exporting']
+
+    def setUp(self):
+        CephFSTestCase.setUp(self)
```
you can change this to `super().setUp()`

zhsgao replied:
> you can change this to `super().setUp()`

OK, I have modified it.

dparmar18 commented on qa/tasks/cephfs/test_dump_export_states.py:
```diff
@@ -0,0 +1,95 @@
+                    if target in info['export_targets']:
+                        return
+        except MaxWhileTries as e:
+            raise RuntimeError('rank {0} has not been added to export target after {1}s'
+                               .format(target, timeout)) from e
```
switch to using f-strings here. would be better readable

zhsgao replied:
> switch to using f-strings here. would be better readable

OK, I have modified it.

dparmar18 commented on qa/tasks/cephfs/test_dump_export_states.py:
```diff
@@ -0,0 +1,95 @@
+        self.status = self.fs.wait_for_daemons()
+
+        self.mount_a.run_shell_payload('mkdir -p test/export')
+
```
You can also add a tearDown() method here(after `setUp()`) calling `super().tearDown()`

zhsgao replied:
> You can also add a tearDown() method here(after `setUp()`) calling `super().tearDown()`

OK, I have added it.

mchangir commented on src/mds/Migrator.h:
```diff
@@ -266,14 +270,42 @@ class Migrator {
     std::map<inodeno_t,std::map<client_t,Capability::Import> > peer_imported;
     MutationRef mut;
     size_t approx_size = 0;
+    // record the start time and time spent of each export state
+    std::map<int, std::pair<utime_t, double> > state_history;
```
just curious to know why `state_history` can't be an `std::array<>` or a `std::vector<>` instead of a `std::map<>` ... since the states advance linearly to the next state ?

is that just to future-proof the code w.r.t. addition of new states ?

zhsgao replied:
> just curious to know why `state_history` can't be an `std::array<>` or a `std::vector<>` instead of a `std::map<>` ... since the states advance linearly to the next state ?
> 
> is that just to future-proof the code w.r.t. addition of new states ?

I think vector can also be used here. I use map because the state value can be used as the key and it's convenient to query the value.

mchangir replied:
well, the state value can be used as an index into the vector<> assuming that EXPORT_NONE isn't represented as a state in the vector<> since its value is -1

vshankar replied:
Do we really need `EXPORT_NONE` state. I think that can be done away with.

zhsgao replied:
> Do we really need `EXPORT_NONE` state. I think that can be done away with.

Yes, I think you are right. I have deleted `EXPORT_NONE`. Thanks!

dparmar18 commented:
This needs to be documented somewhere, if not in docs then in PendingReleaseNotes?

vshankar commented:
@zhsgao Fails api and make check failure:

API test:
```
Undefined Reference

Undefined reference in /usr/include/c++/11/bits/stl_map.h:1216 to `Migrator::EXPORT_DISCOVERING'
[Indication 1](https://jenkins.ceph.com/job/ceph-api/51697/consoleFull#-176439654385e1414f-af06-4588-8fed-ab5891ddc58a)
```

make check:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/CDir.h:553:17: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
    _walk_tree([this, &n](CDir *dir) {
```

zhsgao commented:
> @zhsgao Fails api and make check failure:
> 
> API test:
> 
> ```
> Undefined Reference
> 
> Undefined reference in /usr/include/c++/11/bits/stl_map.h:1216 to `Migrator::EXPORT_DISCOVERING'
> [Indication 1](https://jenkins.ceph.com/job/ceph-api/51697/consoleFull#-176439654385e1414f-af06-4588-8fed-ab5891ddc58a)
> ```
> 
> make check:
> 
> ```
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/CDir.h:553:17: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
>     _walk_tree([this, &n](CDir *dir) {
> ```

I have modified them, thanks!

vshankar commented:
jenkins test make check

vshankar commented:
https://pulpito.ceph.com/vshankar-2023-04-12_10:03:55-fs-wip-vshankar-testing-20230412.053558-testing-default-smithi/

Infra issues with teuthology. Need to rerun tests.

vshankar commented:
test run: https://pulpito.ceph.com/vshankar-2023-04-20_04:50:51-fs-wip-vshankar-testing-20230412.053558-testing-default-smithi/

vshankar commented:
@zhsgao I just realized that the test case source test_dump_export_states is missing a reference in a yaml fragment. I remember seeing it when I reviewed it initially. Did you happen to miss including when pushing an update?

zhsgao commented:
> @zhsgao I just realized that the test case source test_dump_export_states is missing a reference in a yaml fragment. I remember seeing it when I reviewed it initially. Did you happen to miss including when pushing an update?

I remember that I didn't modify any yaml file before in this PR. Which yaml file should I modify or add a new one?

vshankar commented:
> > @zhsgao I just realized that the test case source test_dump_export_states is missing a reference in a yaml fragment. I remember seeing it when I reviewed it initially. Did you happen to miss including when pushing an update?
> 
> I remember that I didn't modify any yaml file before in this PR. Which yaml file should I modify or add a new one?

I recommend moving `TestDumpExportStates` class to `qa/tasks/cephfs/test_exports.py` - no need to create a new yaml fragment then.

zhsgao commented:
> > > @zhsgao I just realized that the test case source test_dump_export_states is missing a reference in a yaml fragment. I remember seeing it when I reviewed it initially. Did you happen to miss including when pushing an update?
> > 
> > 
> > I remember that I didn't modify any yaml file before in this PR. Which yaml file should I modify or add a new one?
> 
> I recommend moving `TestDumpExportStates` class to `qa/tasks/cephfs/test_exports.py` - no need to create a new yaml fragment then.

OK, I've modified it.

vshankar commented:
jenkins retest this please

vshankar commented:
Running tests now: https://pulpito.ceph.com/vshankar-2023-05-09_12:06:46-fs-wip-vshankar-testing-20230509.090020-testing-default-smithi/

vshankar commented:
jenkins test windows

vshankar commented:
@zhsgao The test which caused the MDS segfault is tasks.cephfs.test_exports.TestExports.INFO:tasks.cephfs_test_runner:test_session_race - which actually passed, but left an MDS coredump.

zhsgao commented:
> @zhsgao The test which caused the MDS segfault is tasks.cephfs.test_exports.TestExports.INFO:tasks.cephfs_test_runner:test_session_race - which actually passed, but left an MDS coredump.

I think it's as expected. `mds_kill_import_at` is set to crash the importer to interrupt the export process, and then we test `dump_export_states`.

vshankar commented:
> > @zhsgao The test which caused the MDS segfault is tasks.cephfs.test_exports.TestExports.INFO:tasks.cephfs_test_runner:test_session_race - which actually passed, but left an MDS coredump.
> 
> I think it's as expected. `mds_kill_import_at` is set to crash the importer to interrupt the export process, and then we test `dump_export_states`.

In that case we need to delete the coredump as its done in other tests.

zhsgao commented:
> > > @zhsgao The test which caused the MDS segfault is tasks.cephfs.test_exports.TestExports.INFO:tasks.cephfs_test_runner:test_session_race - which actually passed, but left an MDS coredump.
> > 
> > 
> > I think it's as expected. `mds_kill_import_at` is set to crash the importer to interrupt the export process, and then we test `dump_export_states`.
> 
> In that case we need to delete the coredump as its done in other tests.

OK, I've modified it.

zhsgao commented:
Can we reopen this PR?

vshankar commented:
> Can we reopen this PR? @vshankar

Yes. I've lost a bit of context on this. I'll get back to this.

vshankar commented:
Reopening.

vshankar commented:
Not sure why I'm unable to reopen this PR. @zhsgao maybe push an update?

zhsgao commented:
> Not sure why I'm unable to reopen this PR. @zhsgao maybe push an update?

I have rebased and updated this PR, but it is still closed.
Maybe @batrick knows how to reopen it, he recently reopened another closed PR for me.

zhsgao commented:
> Hovering over the greyed out "Reopen and comment" tab says
> 
> > The mds_export_state branch was force-pushed or recreated
> 
> @zhsgao Suggest to create e fresh PR.

The 'reopen' button is greyed out because I force-pushed after the PR was closed. So I have force-pushed the last commit before the PR was closed. The 'reopen' button should be available now if it worked.

vshankar commented:
> > Hovering over the greyed out "Reopen and comment" tab says
> > > The mds_export_state branch was force-pushed or recreated
> > 
> > 
> > @zhsgao Suggest to create e fresh PR.
> 
> The 'reopen' button is greyed out because I force-pushed after the PR was closed. So I have force-pushed the last commit before the PR was closed. The 'reopen' button should be available now if it worked.

Worked! Thx for the work.

vshankar commented:
@zhsgao please rebase.

zhsgao commented:
> @zhsgao please rebase.

OK, it's done.

vshankar commented:
> > @zhsgao please rebase.
> 
> OK, it's done.

Thank you!

vshankar commented:
jenkins test api

vshankar commented:
This PR is under test in [https://tracker.ceph.com/issues/68578](https://tracker.ceph.com/issues/68578).

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-vshankar-testing-20241016135728-debug
